lucet-wasi/src/hostcalls.rs

```diff
@@ -980,6 +985,211 @@ pub extern "C" fn __wasi_random_get(
     return wasm32::__WASI_ESUCCESS;
 }
 
+nix::ioctl_read_bad!(fionread, nix::libc::FIONREAD, c_int);
+
+fn wasi_clock_to_relative_ns_delay(
+    wasi_clock: host::__wasi_subscription_t___wasi_subscription_u___wasi_subscription_u_clock_t,
+) -> u128 {
+    if wasi_clock.flags != wasm32::__WASI_SUBSCRIPTION_CLOCK_ABSTIME {
+        return wasi_clock.timeout as u128;
+    }
+    let now: u128 = SystemTime::now()
+        .duration_since(SystemTime::UNIX_EPOCH)
+        .expect("Current date is before the epoch")
+        .as_nanos();
+    let deadline = wasi_clock.timeout as u128;
+    deadline.saturating_sub(now)
+}
+
+#[no_mangle]
+pub extern "C" fn __wasi_poll_oneoff(
```

Comment:
In terms of text, this is a beast of a function, but it doesn't actually seem all that complicated now that I've read through it. I think that if you added comments explaining each of the major steps (e.g., decoding subscriptions, splitting subscription types, converting to pollfd, polling, gathering results), it would be much easier to understand at first glance. Given that this is largely a pipeline rather than repeated operations on the same values, it would probably also be suitable for splitting some of these steps into helper functions and having a (relatively) simple driver function for the body of the hostcall.

Reply:
done